pm/helpers.go

```diff
@@ -1,27 +1,61 @@
 package pm
 
 import (
-	"fmt"
+	"math/big"
 	"math/rand"
 
+	"github.com/ethereum/go-ethereum/accounts"
 	ethcommon "github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/pkg/errors"
+)
+
+var (
+	secp256k1N, _  = new(big.Int).SetString("fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141", 16)
+	secp256k1halfN = new(big.Int).Div(secp256k1N, big.NewInt(2))
 )
 
 // VerifySig verifies that a ETH ECDSA signature over a given message
 // is produced by a given ETH address
 //
 // TODO refactor to a package that both eth and pm can import
 func VerifySig(addr ethcommon.Address, msg, sig []byte) bool {
-	personalMsg := fmt.Sprintf("\x19Ethereum Signed Message:\n%d%s", 32, msg)
-	personalHash := crypto.Keccak256([]byte(personalMsg))
-
-	pubkey, err := crypto.SigToPub(personalHash, sig)
+	recovered, err := ecrecover(msg, sig)
 	if err != nil {
 		return false
 	}
 
-	return crypto.PubkeyToAddress(*pubkey) == addr
+	return recovered == addr
+}
+
+func ecrecover(msg, sig []byte) (ethcommon.Address, error) {
```

Comment:
I guess we're using a custom `ecrecover` implementation here because https://godoc.org/github.com/ethereum/go-ethereum/crypto#Ecrecover doesn't check whether the recovery id is `27 || 28` ?

Reply:
`crypto.SigToPub()` expects a signature with v = 0/1. Furthermore, the non-cgo version of `crypto.SigToPub()` does not check that the signature's s-value is in the lower range. So, in order to make sure that local signature verification matches on-chain signature verification I added checks in `pm.ecrecover()` that correspond to the checks in the on-chain signature verification library (i.e. invalid signature length, invalid s-value, invalid v-value)

Chose to use `crypto.SigToPub()` instead of `crypto.Ecrecover()` in this function because it seemed more straightforward to convert the pub key returned by `crypto.SigToPub()` into a `common.Address` using `crypto.PubkeyToAddress()` - `crypto.Ecrecover()` returns a `[]byte` which would need to be converted into a pub key before finally converting into `common.Address`.